### PR TITLE
add api for fetching federated query configs

### PR DESCRIPTION
**Pull request by tiana528**

### CODE REVIEW

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
+package com.treasuredata.client.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.base.Objects;
```
Comment (yuokada) [inline src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java:6]: In `td-client-java`, we have abolished usage of `com.google.common.base.Objects`,

Could you rewrite the class without using `com.google.common.base.Objects`?

Comment (tiana528) [reply src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java]: Sure! Updated

File: src/main/java/com/treasuredata/client/deserialize/FederatedQueryConfigSettingsDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import java.io.IOException;
+
+public class FederatedQueryConfigSettingsDeserializer extends JsonDeserializer<String>
```
Comment (yuokada) [inline src/main/java/com/treasuredata/client/deserialize/FederatedQueryConfigSettingsDeserializer.java:11]: Could you reformat the file with 4 spaces indent style?

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
+package com.treasuredata.client.model;
+
+import java.util.Objects;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
+public class TDFederatedQueryConfig
```
Comment (yuokada) [inline src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java:9]: Could you reformat the file with 4 spaces indent style?

File: src/test/java/com/treasuredata/client/TestTDClient.java [hunk 4]
```diff
@@ -1958,6 +1959,50 @@ public void testImportBytesWithId()
...
+        server.enqueue(new MockResponse().setBody(federatedQueryConfigs));
+
+        List<TDFederatedQueryConfig> result = client.getFederatedQueryConfigs();
+
+        List<TDFederatedQueryConfig> expected = Arrays.asList(
+            new TDFederatedQueryConfig(12345, "custom_type", 67890, 54321, "test_name", 112233, "2023-06-01T00:00:00Z", "2023-06-15T00:00:00Z", "{\"account_name\":\"snowflake_account\",\"auth_method\":\"password\",\"private_key\":\"encrypted_key\",\"passphrase\":\"secret_phrase\",\"options\":{\"key1\":\"val1\",\"key2\":\"val2\"},\"user\":\"snowflake_user\",\"password\":\"snowflake_password\",\"database\":\"sample_database\"}"),
+            new TDFederatedQueryConfig(67890, "another_type", 12345, 98765, "another_test_name", 445566, "2023-07-01T00:00:00Z", "2023-07-15T00:00:00Z", "{\"account_name\":\"another_snowflake_account\",\"auth_method\":\"oauth\",\"private_key\":\"another_encrypted_key\",\"passphrase\":\"another_secret_phrase\",\"options\":{\"keyA\":\"valA\",\"keyB\":\"valB\"},\"user\":\"another_snowflake_user\",\"password\":\"another_snowflake_password\",\"database\":\"another_sample_database\"}")
+        );
+        assertEquals(result, expected);
```
Comment (yuokada) [inline src/test/java/com/treasuredata/client/TestTDClient.java:2003]: `Assertions.assertIterableEquals` should be used.
```suggestion
        assertIterableEquals(result, expected);
```

File: src/test/java/com/treasuredata/client/TestTDClient.java [hunk 5]
```diff
@@ -1958,6 +1959,50 @@ public void testImportBytesWithId()
...
+        server.enqueue(new MockResponse().setBody(federatedQueryConfigs));
+
+        List<TDFederatedQueryConfig> result = client.getFederatedQueryConfigs();
+
+        List<TDFederatedQueryConfig> expected = Arrays.asList(new TDFederatedQueryConfig(12345, "custom_type", 67890, 54321, "test_name", 112233, "2023-06-01T00:00:00Z", "2023-06-15T00:00:00Z", "{\"account_name\":\"snowflake_account\",\"auth_method\":\"password\",\"private_key\":\"encrypted_key\",\"passphrase\":\"secret_phrase\",\"options\":{\"key1\":\"val1\",\"key2\":\"val2\"},\"user\":\"snowflake_user\",\"password\":\"snowflake_password\",\"database\":\"sample_database\"}"));
+        assertEquals(result, expected);
```
Comment (yuokada) [inline src/test/java/com/treasuredata/client/TestTDClient.java:1985]: `Assertions.assertIterableEquals` should be used.
```suggestion
        assertIterableEquals(result, expected);
```

File: src/test/java/com/treasuredata/client/TestTDClient.java [hunk 6]
```diff
@@ -1958,6 +1959,50 @@ public void testImportBytesWithId()
...
+
+        assertTrue(result.isEmpty());
+    }
+
+    @Test
+    public void testGetFederatedQueryConfigsWhenOnlyOne()
+        throws Exception
+    {
+        client = mockClient();
+        final String federatedQueryConfigs = "[{\"id\":12345,\"type\":\"custom_type\",\"user_id\":67890,\"account_id\":54321,\"name\":\"test_name\",\"connection_id\":112233,\"created_at\":\"2023-06-01T00:00:00Z\",\"updated_at\":\"2023-06-15T00:00:00Z\",\"settings\":{\"account_name\":\"snowflake_account\",\"auth_method\":\"password\",\"private_key\":\"encrypted_key\",\"passphrase\":\"secret_phrase\",\"options\":{\"key1\":\"val1\",\"key2\":\"val2\"},\"user\":\"snowflake_user\",\"password\":\"snowflake_password\",\"database\":\"sample_database\"}}]";
```
Comment (yuokada) [inline src/test/java/com/treasuredata/client/TestTDClient.java:1979]: Could you add some line feeds to reformat the json string?

File: src/test/java/com/treasuredata/client/TestTDClient.java [hunk 7]
```diff
@@ -1958,6 +1959,50 @@ public void testImportBytesWithId()
...
+        assertEquals(result, expected);
+    }
+
+    @Test
+    public void testGetFederatedQueryConfigsWhenMultiple()
+        throws Exception
+    {
+        client = mockClient();
+        final String federatedQueryConfigs = "[{\"id\":12345,\"type\":\"custom_type\",\"user_id\":67890,\"account_id\":54321,\"name\":\"test_name\",\"connection_id\":112233,\"created_at\":\"2023-06-01T00:00:00Z\",\"updated_at\":\"2023-06-15T00:00:00Z\",\"settings\":{\"account_name\":\"snowflake_account\",\"auth_method\":\"password\",\"private_key\":\"encrypted_key\",\"passphrase\":\"secret_phrase\",\"options\":{\"key1\":\"val1\",\"key2\":\"val2\"},\"user\":\"snowflake_user\",\"password\":\"snowflake_password\",\"database\":\"sample_database\"}},"
```
Comment (yuokada) [inline src/test/java/com/treasuredata/client/TestTDClient.java:1993]: Could you add some line feeds to reformat the json string?

File: src/test/java/com/treasuredata/client/TestTDClient.java [hunk 8]
```diff
@@ -1958,6 +1959,50 @@ public void testImportBytesWithId()
         assertEquals(result.getUniqueId(), "4288048cf8f811e88b560a87157ac806");
     }
 
+    @Test
+    public void testGetFederatedQueryConfigsWhenEmpty()
```
Comment (yuokada) [inline src/test/java/com/treasuredata/client/TestTDClient.java:1963]: Could you add test cases that the endpoint return 4xx status code?
The endpoint returns `401 and 403`.

Comment (tiana528) [reply src/test/java/com/treasuredata/client/TestTDClient.java:1964]: It is not very convenient to simulate 401 and 403. I added a test case of 401, while it depends on the internal behaviour. 
To support 401 and 403 in the test case, we need to configure the account of staging-aws used by the unit tests in a way to trigger 401 and 403 (Some junit tests in this class actually sends real request to staging-aws api3), which I feel might be a little too much.

Comment (yuokada) [reply src/test/java/com/treasuredata/client/TestTDClient.java:1964]: sure, I see.

**Review by yuokada — APPROVED** [review]
LGTM

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        int result = id;
+        result = 31 * result + Objects.hashCode(type);
+        result = 31 * result + userId;
+        result = 31 * result + accountId;
+        result = 31 * result + Objects.hashCode(name);
+        result = 31 * result + connectionId;
+        result = 31 * result + Objects.hashCode(createdAt);
+        result = 31 * result + Objects.hashCode(updatedAt);
+        result = 31 * result + Objects.hashCode(settings);
+        return result;
```
Comment (miniway) [inline src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java:133]: why not `Objects.hash(id, type, userId, accountId, ...)`

Comment (yuokada) [reply src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java:133]: In my understanding, that code is automatically generated by IntelliJ or from any template.

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    private final String type;
+    private final int userId;
+    private final int accountId;
+    private final String name;
+    private final int connectionId;
+    private final String createdAt;
+    private final String updatedAt;
+
+    @JsonDeserialize(using = com.treasuredata.client.deserialize.FederatedQueryConfigSettingsDeserializer.class)
+    private final String settings;
```
Comment (miniway) [inline src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java:21]: If this is anonymous json serialized, `Map<String, Object>` should be enough

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java (hunk 9 above)
Comment (tiana528) [reply src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java:133]: Yes, it is generated by IntelliJ automatically.

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java (hunk 10 above)
Comment (tiana528) [reply src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java]: Updated like that.

File: src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java (hunk 9 above)
Comment (tiana528) [reply src/main/java/com/treasuredata/client/model/TDFederatedQueryConfig.java]: Changed to Objects.hash(id, type, userId, accountId, ...)

**Review by miniway — APPROVED** [review]
LGTM

**Comment by tiana528** [comment]
Thank you!

**Pull request merged**